electricitymap/contrib/libs/models/datapoints.py

```diff
@@ -0,0 +1,427 @@
+        if v > 500000:
+            raise ValueError(f"Generation is implausibly high, above 500GW: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zoneKey: ZoneKey,
+        datetime: datetime,
+        source: str,
+        value: float,
+        forecasted: bool = False,
+    ) -> Optional["Generation"]:
+        try:
+            return Generation(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                value=value,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating generation Event {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "generation": self.value,
+            "source": self.source,
+        }
+
+
+class ProductionBreakdown(Event):
+    production: ProductionMix
+    storage: Optional[StorageMix] = None
+
+    @validator("production", "storage")
+    def _validate_mix(cls, v):
+        if v is not None:
+            if all(value is None for value in v.dict().values()):
+                raise ValueError("Mix is completely empty")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zoneKey: ZoneKey,
+        datetime: datetime,
+        source: str,
+        production: ProductionMix,
+        storage: Optional[StorageMix] = None,
+        forecasted: bool = False,
+    ) -> Optional["ProductionBreakdown"]:
+        try:
+            # Correct negative production values.
+            for key, value in production.dict().items():
+                if value is not None and value < 0:
+                    production.__setattr__(key, None)
+                    logger.warning(
+                        f"Production value for {key} is negative.\
+                              This value is set to None."
+                    )
+            return ProductionBreakdown(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                production=production,
+                storage=storage,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating production breakdown Event {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "production": self.production.dict(exclude_none=True),
+            "storage": self.storage.dict(exclude_none=True) if self.storage else None,
+            "source": self.source,
+        }
+
+
+class Consumption(Event):
+    consumption: float
+
+    @validator("consumption")
+    def _validate_consumption(cls, v: float):
+        if v < 0:
+            raise ValueError(f"Consumption cannot be negative: {v}")
+        if v > 500000:
```

Comment:
Preferably it would be nice if we could specify this value per zone instead of having one global static value. Perhaps something for v2?

Reply:
Yes, this one is just a baseline check to make sure values are not crazy. A per zone layer will be added at a later stage, as part of the backend.